Add Zip/UnZip Tests

mjkkirschner opened the pull request:
### Purpose

https://jira.autodesk.com/browse/QNTM-4781

updates Greg.dll (packageManagerClient) to use system.io.compression for both zipping and unzipping - building on @Dewb 's work.

adds two tests that use packageManagerClient zipping and unzipping functions.

### Declarations

TODO:
- [x] remove ionic.zip references from this repo if possible.

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@QilongTang 
@alfarok 
@scottmitchell 

### FYIs

@smangarole

mjkkirschner commented:
builds are failing until we upload a matching nuget

QilongTang commented on src/DynamoCore/packages.config:
```diff
@@ -1,6 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="Greg" version="1.0.6176.18754" targetFramework="net45" />
+  <package id="Greg" version="1.1.6905.23659" targetFramework="net47" />
```
Oh, I was not aware the .Net change is merged. How do we publish the nuget of Greg?

mjkkirschner replied:
its not merged, I pushed the nugets only to my local repo. I will push the nugets manually once this PR https://github.com/DynamoDS/PackageManagerClient/pull/20 gets LGTM

mjkkirschner replied:
the 47 here is automatically added - it is just the framework version of the project referencing this dll - it lets nuget know to prefer that version if it exists, but it doesnt exist.

QilongTang replied:
Make sense

QilongTang commented on test/Libraries/PackageManagerTests/PackageUploadBuilderTests.cs:
```diff
@@ -29,6 +32,68 @@ private IPackageUploadBuilder BigPackageUploadBuilderMock()
+            public void FileCompressorZipsSamplePackageDirectoryToValidArchive()
+        {
+            var executingLocation = new DirectoryInfo(Assembly.GetExecutingAssembly().Location);
+            var testDirectory = Path.Combine(executingLocation.Parent.Parent.Parent.Parent.FullName, "test");
```
where is it pointing?

mjkkirschner replied:
talked about this with @QilongTang yesterday, this is pointing back up the /test/pkgs folder.

alfarok commented on src/DynamoCore/DynamoCore.csproj:
```diff
@@ -78,9 +78,8 @@ limitations under the License.
-      <HintPath>..\packages\Greg.1.0.6176.18754\lib\net45\Greg.dll</HintPath>
-      <Private>True</Private>
+    <Reference Include="Greg, Version=1.1.6905.23659, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>..\packages\Greg.1.1.6905.23659\lib\net45\Greg.dll</HintPath>
```
@mjkkirschner should this package be in a `\lib\net47\Greg.dll` location now?

mjkkirschner replied:
nope, we have not updated greg.dll to net 47 yet.

alfarok commented on test/pkgs/the-Saurus/extra/thesaurus_ViewExtensionDefinition.xml:
```diff
@@ -0,0 +1,4 @@
+﻿<ViewExtensionDefinition>
```
this is minor but in the future should we setup a `test/extensions` folder for extensions even though they are technically still packages just for clarity?

mjkkirschner replied:
If you would like to update the tests 😉 .

mjkkirschner commented:
@kronz - I have modified https://github.com/DynamoDS/Dynamo/blob/5a5d625b761453788f42392187be634fb9361ba1/doc/distrib/License.rtf

as I've removed ionic.zip nugets from the repo - do we need to run this past James?

mjkkirschner commented:
when it is indexed I will update these refs to https://www.nuget.org/packages/Greg/1.1.6907.32055

kronz commented:
@mjkkirschner if there are libraries that have been removed, we just need to check in our about box and any other place that we might have refererenced it and remove the reference. no need to ask James if we are just looking at a subtraction.  Thanks

mjkkirschner commented on test/Libraries/PackageManagerTests/PackageUploadBuilderTests.cs:
```diff
@@ -29,6 +32,68 @@ private IPackageUploadBuilder BigPackageUploadBuilderMock()
+
+            var compressor = new MutatingFileCompressor();
+            var zipPath = compressor.Zip(packageDir);
+            Assert.AreEqual(1306268, zipPath.Length);
```
this test zips some dlls which caused this issue and other zipping problems and makes sure the zip is the correct size and unzips afterwards to the correct number of files.

scottmitchell commented on test/Libraries/PackageManagerTests/PackageUploadBuilderTests.cs:
```diff
@@ -29,6 +32,68 @@ private IPackageUploadBuilder BigPackageUploadBuilderMock()
+            var unzipPath = Greg.Utility.FileUtilities.UnZip(zipPath.Name);
+
+            Assert.AreEqual(preZipDiskSize, new DirectoryInfo(unzipPath).GetFiles().Sum(x => x.Length));
+            Assert.AreEqual(preZipfileCount, new DirectoryInfo(unzipPath).GetFiles().Length);
```
Not super familiar with what happens behind the scenes when zipping/unzipping, but is it at all likely that the size and number of files could be consistent yet still have been corrupted somehow? Would it be insane to check individual bytes?

mjkkirschner replied:
I think in general if we get this far in the test we're good, previously this bug would have thrown an exception.

Dewb replied:
Late to the party here, but I think it would be reasonable to have a test that assembles a folder, zips the folder, unzips the zip into another folder, and compares the contents to ensure they're identical.

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/how-to-compare-the-contents-of-two-folders-linq

mjkkirschner replied:
@Dewb ended up doing that in another PR using md5 hashes for each file.
https://github.com/DynamoDS/Dynamo/pull/9281

scottmitchell commented:
LGTM!

jnealb commented:
@Dewb @scottmitchell I suggested just that.
